### PR TITLE
Fix T5Attention shape mismatch under Tensor Parallelism

### PR DESCRIPTION
**Pull request by aws-zhanxun**
# What does this PR do?

`T5Attention.forward` hard-codes `n_heads` and `inner_dim` in `view()` calls. When using PyTorch Tensor Parallelism, `ColwiseParallel` shards the q/k/v projection output dim from `inner_dim` to `inner_dim / tp_size`, but `n_heads` remains unchanged, causing a shape mismatch.

For example, with `n_heads=128`, `d_kv=128`, and `tp_size=4`, for an input with `seq_length=512`:

```python
# After ColwiseParallel, q output shape: [B, 512, 16384 / 4] = [B, 512, 4096]

# Original code:
q.view(B, -1, 128, 128)
# -1 infers seq_length = 512 * 4096 / (128 * 128) = 128 ≠ 512
```

For comparison, Llama-style attention already uses `view(batch, seq, -1, head_dim)` where `-1` is on the `num_heads` dim, making it TP-transparent. T5 puts `-1` on the `seq_length` dim instead.

This PR moves the `-1` from the `seq_length` dim to the `num_heads` dim, so that `num_heads` is inferred from the actual tensor shape rather than the config value.

## Reproducer

<details>
<summary>repro_t5_tp.py (requires >= 2 GPUs)</summary>

```python
"""
Minimal reproducer: T5Attention view() shape mismatch under Tensor Parallelism.

Uses real T5 model + PyTorch TP APIs. Requires >= 2 GPUs.
Run with: torchrun --nproc_per_node=2 repro_t5_tp.py
"""

import os
import torch
import torch.distributed as dist
from torch.distributed.tensor.parallel import ColwiseParallel, RowwiseParallel, parallelize_module
from transformers import T5Config, T5ForConditionalGeneration


def main():
    dist.init_process_group("nccl")
    rank = dist.get_rank()
    world_size = dist.get_world_size()
    device = torch.device(f"cuda:{rank}")
    torch.cuda.set_device(device)

    # Small T5 config for quick repro
    config = T5Config(
        vocab_size=32128,
        d_model=512,
        d_kv=64,
        d_ff=2048,
        num_heads=8,
        num_layers=2,
        num_decoder_layers=2,
    )
    model = T5ForConditionalGeneration(config).to(device)

    # Apply ColwiseParallel / RowwiseParallel to attention layers
    # (same pattern as PyTorch TP would apply)
    for name, module in model.named_modules():
        if module.__class__.__name__ == "T5Attention":
            plan = {
                "q": ColwiseParallel(),
                "k": ColwiseParallel(),
                "v": ColwiseParallel(),
                "o": RowwiseParallel(),
            }
            if hasattr(module, "relative_attention_bias"):
                plan["relative_attention_bias"] = ColwiseParallel()
            mesh = torch.distributed.device_mesh.init_device_mesh("cuda", (world_size,))
            parallelize_module(module, mesh, plan)
            if rank == 0:
                print(f"  Parallelized: {name}")

    # Forward pass
    input_ids = torch.randint(0, config.vocab_size, (1, 32), device=device)
    decoder_input_ids = torch.randint(0, config.vocab_size, (1, 16), device=device)

    if rank == 0:
        print("\nRunning forward pass...")

    try:
        with torch.no_grad():
            output = model(input_ids=input_ids, decoder_input_ids=decoder_input_ids)
        if rank == 0:
            print(f"SUCCESS — output logits shape: {output.logits.shape}")
    except RuntimeError as e:
        if rank == 0:
            print(f"FAILED — {e}")

    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

</details>

**Before fix** (transformers 5.3.0, 2x NVIDIA H200, TP=2):
```
FAILED — The size of tensor a (16) must match the size of tensor b (32) at non-singleton dimension 2
```

**After fix**:
```
SUCCESS — output logits shape: torch.Size([1, 16, 32128])
```

- [x] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @cyrilvallez for TP I think?

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -274,7 +274,7 @@ def forward(
         is_cross_attention = key_value_states is not None
 
         query_states = self.q(hidden_states)
-        query_states = query_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
+        query_states = query_states.view(batch_size, seq_length, -1, self.key_value_proj_dim).transpose(1, 2)
```
Comment (vasqu) [inline src/transformers/models/t5/modeling_t5.py]: Everything is correct and a lot of older models are affected by this pattern. Can we follow more recent shape styles, e.g. https://github.com/vasqu/transformers/blob/943bce340806ae13abbb78405d09c1ae4bbf3486/src/transformers/models/t5/modeling_t5.py#L321-L322 (I worked on SDPA support for T5 at some point)

A more modern version would be in Bert Cross attention instead https://github.com/huggingface/transformers/blob/8213e0d920d52cb00dcade16b6d1f6e952ac0a8c/src/transformers/models/bert/modeling_bert.py#L242-L284

**Review by vasqu — APPROVED** [review]
Careful approval because it's correct but I would like to follow more modern versions where we define the shape everywhere before as tuples. 

You will need to use `make fix-repo` to apply the changes to other models that inherit from T5 (which is why CI is red).

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -273,8 +273,10 @@ def forward(
         # if key_value_states are provided this layer is used as a cross-attention layer for the decoder
         is_cross_attention = key_value_states is not None
 
+        q_input_shape = (batch_size, seq_length, -1, self.key_value_proj_dim)
+
         query_states = self.q(hidden_states)
-        query_states = query_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
+        query_states = query_states.view(*q_input_shape).transpose(1, 2)
```
Comment (vasqu) [inline src/transformers/models/t5/modeling_t5.py]:
```suggestion
        q_input_shape = (batch_size, seq_length, -1, self.key_value_proj_dim)
        query_states = self.q(hidden_states).view(*q_input_shape).transpose(1, 2)
```

File: src/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -294,10 +296,11 @@ def forward(
             key_states = curr_past_key_values.layers[self.layer_idx].keys
             value_states = curr_past_key_values.layers[self.layer_idx].values
         else:
+            kv_input_shape = (batch_size, current_states.shape[1], -1, self.key_value_proj_dim)
             key_states = self.k(current_states)
             value_states = self.v(current_states)
-            key_states = key_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
-            value_states = value_states.view(batch_size, -1, self.n_heads, self.key_value_proj_dim).transpose(1, 2)
+            key_states = key_states.view(*kv_input_shape).transpose(1, 2)
+            value_states = value_states.view(*kv_input_shape).transpose(1, 2)
```
Comment (vasqu) [inline src/transformers/models/t5/modeling_t5.py]:
```suggestion
            kv_input_shape = (batch_size, current_states.shape[1], -1, self.key_value_proj_dim)
            key_states = self.k(current_states).view(*kv_input_shape).transpose(1, 2)
            value_states = self.v(current_states).view(*kv_input_shape).transpose(1, 2)
```

**Review by vasqu — APPROVED** [review]
Just a small nit on directly doing everything on one line re projection + view

**Comment by vasqu** [comment]
run-slow: longt5, mt5, pop2piano, switch_transformers, t5, udop

**Comment by vasqu** [comment]
Just checking with run-slow but LGTM

**Comment by vasqu** [comment]
Thanks a lot for iterating and the contribution :hugs: @aws-zhanxun
